dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/BitBucketSourceCodeRepo.java

```diff
@@ -212,24 +247,55 @@ public void updateReferenceType(String repositoryId, Version version) {
         RefsApi refsApi = new RefsApi(apiClient);
         // There isn't exactly a single Bitbucket endpoint to get a version which then allows us to determine if it's a branch or tag.
         // This code checks two endpoints (branches and tags) to see if it belongs in which.
-        try {
-            refsApi.repositoriesUsernameRepoSlugRefsBranchesNameGet(workspace, name, repoSlug);
-            version.setReferenceType(Version.ReferenceType.BRANCH);
-            return;
-        } catch (ApiException e) {
-            LOG.warn(gitUsername + ": apiexception on reading branches" + e.getMessage(), e);
-            // this is not so critical to warrant a http error code
+        boolean rateLimited;
+        do {
+            try {
+                refsApi.repositoriesUsernameRepoSlugRefsBranchesNameGet(workspace, name, repoSlug);
+                version.setReferenceType(Version.ReferenceType.BRANCH);
+                return;
+            } catch (ApiException e) {
+                rateLimited = isRateLimited(false, e, "reading branches");
+            }
+
+            try {
+                refsApi.repositoriesUsernameRepoSlugRefsTagsNameGet(workspace, name, repoSlug);
+                version.setReferenceType(Version.ReferenceType.TAG);
+                return;
+            } catch (ApiException e) {
+                // this is not so critical to warrant a http error code
+                rateLimited = isRateLimited(rateLimited, e, "reading tags");
+            }
+        } while (rateLimited);
+
+        throw new CustomWebApplicationException(name + " is not a Bitbucket branch or tag in " + repositoryId, HttpStatus.SC_INTERNAL_SERVER_ERROR);
+    }
+
+    /**
+     * determine whether a request is rate-limited, output logging mesages, and potentially sleep past rate-limiting
+     * @param rateLimited
+     * @param e
+     * @param message
+     * @return
+     */
+    private boolean isRateLimited(final boolean rateLimited, ApiException e, String message) {
+        if (LOG.isWarnEnabled()) {
+            LOG.warn("%s: apiexception on %s%s".formatted(gitUsername, message, e.getMessage()), e);
         }
+        // this is not so critical to warrant a http error code
+        boolean newlyRateLimited = false;
+        if (e.getCode() == Status.TOO_MANY_REQUESTS.getStatusCode()) {
```

Comment:
Will the code that uses `isRateLimited` continue to loop for as long as BitBucket returns `TOO_MANY_REQUESTS`?  If so, we might want to limit the number of retries so that if BitBucket gets "stuck" in that state (for whatever reason), we eventually abort.

Reply:
Yes, not sure how the code could tell what was too many retries vs. we legitimately having too many jobs competing for rate limited requests though. Think a partial solution for this is `we could have someone fork/use a queue https://circleci.com/developer/orbs/orb/eddiewebb/queue to make sure bitbucket jobs do not compete with each other, trading-off quota`

Reply:
Alternatively, there's a max 10 minute no output limit https://support.circleci.com/hc/en-us/articles/360007188574-Build-has-Hit-Timeout-Limit

Reply:
Conversation in https://github.com/dockstore/dockstore/pull/5313#discussion_r1072803570